amethyst_assets/src/storage.rs

```diff
@@ -112,46 +118,133 @@ impl<A: Asset> AssetStorage<A> {
+                            .map_err(|e| AssetError::new(name, format, e))?;
+
+                        let id = handle.id();
+                        bitset.add(id);
+                        handles.push(handle.clone());
+
+                        // NOTE: the loader has to ensure that a handle will be used
+                        // together with a `Data` only once.
+                        unsafe {
+                            assets.insert(id, asset);
+                        }
+
+                        (reload_obj, handle)
+                    }
+                    Processed::HotReload {
+                        data,
+                        format,
+                        handle,
+                        name,
+                        old_reload,
+                    } => {
+                        let (asset, reload_obj) = match data.map(
+                            |FormatValue { data, reload }| (data, reload),
+                        ).and_then(|(d, rel)| f(d).map(|a| (a, rel)))
+                            .map_err(|e| AssetError::new(name, format, e))
+                        {
+                            Ok(x) => x,
+                            Err(e) => {
+                                eprintln!("Failed to hot-reload: {}", e);
+
+                                reloads.push((handle.downgrade(), old_reload));
+
+                                return Ok(());
+                            }
+                        };
+
+                        let id = handle.id();
+                        assert!(bitset.contains(id));
+                        unsafe {
+                            let old = assets.get_mut(id);
+                            *old = asset;
+                        }
+
+                        (reload_obj, handle)
+                    }
+                };
+
+                // Add the reload obj if it is `Some`.
+                if let Some(reload_obj) = reload_obj {
+                    reloads.push((handle.downgrade(), reload_obj));
                 }
 
                 Ok(())
             });
         }
 
-        while let Some(i) = self.handles.iter().position(Handle::is_unused) {
-            let old = self.handles.swap_remove(i);
-            let id = i as u32;
+        while let Some(i) = self.handles.iter().position(Handle::is_unique) {
+            let handle = self.handles.swap_remove(i);
+            let id = handle.id();
             unsafe {
                 self.assets.remove(id);
             }
             self.bitset.remove(id);
-            self.unused_handles.push(old);
+
+            // Can't reuse old handle here, because otherwise weak handles would still be valid.
+            // TODO: maybe just store u32?
+            self.unused_handles.push(Handle {
+                id: Arc::new(id),
+                marker: PhantomData,
+            });
+        }
+
+        // Reload every seconds
+        // TODO: more configuration
```

Comment:
You can add more details to `TODO`.
Configure timeout.
Configure automatic or only manual with special method in `AssetStorage`.

Reply:
I will add such a structure in this PR.

Reply:
Cool!